<a href="https://colab.research.google.com/github/shaheerzubery/DataCleaning/blob/main/Imbalance_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### IMBALANCE DATASET - UnderSampling and Oversampling 

## Impot libraries

In [2]:
import numpy as np
import pandas as pd
from sklearn import preprocessing
from collections import Counter

## Import data

In [3]:
df = pd.read_csv('/content/Imbalance_dataset.csv')

In [4]:
df.shape

(4187, 2)

In [7]:
df.head()

Serial Number  Distibutor
0              1           0
1              2           0
2              3           0
3              4           0
4              5           0

we have 2 columns and 4187 records

## Check for data Imbalance

In [8]:
print(sorted((Counter(df['Distibutor']).items())))

[(0, 833), (1, 3354)]


So we have 833 Zeros and 3354 Ones

In [9]:
pip install imblearn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [10]:
x = df
y = df['Distibutor']

In [11]:
from imblearn import under_sampling, over_sampling

In [12]:
from imblearn.over_sampling import RandomOverSampler
ros= RandomOverSampler(random_state = 0)
x_resampled, y_resampled = ros.fit_resample(x,y)
print(sorted((Counter(y_resampled).items())),y_resampled.shape)

[(0, 3354), (1, 3354)] (6708,)


Now our data is over sampled

In [13]:
from imblearn.under_sampling import RandomUnderSampler
ros= RandomUnderSampler(random_state = 0)
x_resampled, y_resampled = ros.fit_resample(x,y)
print(sorted((Counter(y_resampled).items())),y_resampled.shape)

[(0, 833), (1, 833)] (1666,)


Now the data is Undersampled

Undersample is recommended when there is large amount of data